package.json

```diff
@@ -76,6 +77,14 @@
     "pirates": "^4.0.1",
     "ts-interface-checker": "^0.1.9"
   },
+  "peerDependencies": {
+    "ts-node": ">=10.9.1"
+  },
+  "peerDependenciesMeta": {
+    "ts-node": {
+      "optional": true
+    }
+  },
```

Comment:
@cspotcode I've been trying to wrap my head around the right way to think about peer dependencies and compatibility expectations here and wanted to get your thoughts and any recommendations.

I based these package.json lines off of the peer dependency you have on SWC; I hadn't previously seen `peerDependenciesMeta` and `optional`, but definitely seems like the right approach for this sort of integration.

The `>=` here is certainly convenient for me because Sucrase won't need to be updated at every ts-node semver-major release, but the worry, of course, is that ts-node breaking changes will break the plugin API. I suppose for the Sucrase case specifically, you'll presumably be careful to not break `sucrase: true`, so we can coordinate whatever migration path makes sense as necessary. But thinking more generally, I could see three ways to think about it:
* The `>=` is correct because ts-node can commit to not breaking old plugins, not even in semver-major releases.
* ts-node may have breaking changes to the plugin interface, but the `>=` is fine because it still provides the guarantee to users that their tools will only break while doing a semver-major upgrade to *some* tool (`ts-node` *or* `sucrase`). The main goal is to avoid a *surprise* break while doing a minor upgrade or regenerating a lockfile, and if `ts-node` breaks the plugin interface, users might need to upgrade `sucrase` or wait for a new version.
* The `>=` is incorrect, and it would be better to list the major versions that I know are compatible and try to keep them updated in a timely way.

Sounds like maybe the second bullet is the most reasonable way to think about it.

If I'm following the advice at https://github.com/semver/semver/issues/502#issuecomment-466501843 , any increase to the minimum ts-node version here is semver-major for Sucrase. It certainly makes sense; re-generating your lockfile (or doing a minor upgrade to one tool) should never break your tools. Some consequences to think about in this case:
* If I release right now and then you later add new convenience features like `isTS` or `isJSX` in a new version, I won't be able to rely on them until the next Sucrase semver-major release, since I need to still support ts-node 10.9.1. Other types of changes like optional diagnostics in the response would be fine for me to add, as long as ts-node doesn't break on extraneous fields now (and I see that it doesn't).
* If ts-node releases a breaking change without providing a path for me to maintain support for old and new ts-node versions in the same plugin, that puts me in a hard spot because it would mean that the only way for me to fix compatibility is with a semver-major upgrade (which I'd generally prefer to avoid). One ad hoc way to avoid this problem is to pick a different function name, e.g. the old API is `create` and the new one is `createTranspiler` and I include both for compatibility for a while and then eventually drop the old `create` in a breaking change. From https://github.com/TypeStrong/ts-node/issues/1851, sounds like you also had some ideas around API version to coordinate this.

Another approach to all of this is to have another package `@sucrase/ts-node-plugin` that people install in addition to the other two. That would make major version bumps less of a big deal since it's just the integration with the semver-major change, but I think including it in the core `sucrase` package certainly seems nicest from a DX perspective.

Maybe I'm overthinking these things and there won't be a need for significant breaking changes, just thinking through how it would all work. Like I mentioned, I'm sure we can coordinate since you wouldn't want `sucrase: true` to break anyway.

My impression from the conversation so far is that breaking changes won't actually be necessary and the existing interface makes additive changes pretty easy, but would be good to know if you have a significant breaking change in mind. To be clear, I'd be fine shipping this as-is without the API interface improvements and then picking them up whenever I do a semver-major upgrade in the future, or I'd also be fine waiting for improvements or breaking changes to the interface if that makes the most sense.

Reply:
After some more thought and testing, I'm now thinking that I won't add even an optional peer dependency, since it could break npm users where a project has sucrase and ts-node installed independently without plans to use the plugin. I can always add docs as needed or possibly other approaches to ensure compatibility, but for now will keep it simple.